### DISCUSSION TITLE
ConvexPolyhedron with imported geometry can't collide against each other ?

### DISCUSSION THREAD

**Discussion by Faber-smythe** [pmndrs/use-cannon#135] [Q&A] [answered]
Hi there ! 
First I'm a newbie so tell me right away if I'm asking in the wrong place or didn't get something basic.

I'm importing basic meshes from Blender and throwing them on a plane. But it crashes as soon as they collide against each other. I don't get why. 
Here is a straight forward sandbox to reproduce my case. Has anybody a clue as to what could be the problem ?

-------------------------------

It might not be related, but it was pointed to me that in cannon.js [ConvexPolyhedron class doesn't accept coplanar triangles](https://discordapp.com/channels/740090768164651008/757848836294311957/783302377578954762) :
"No polygons may be coplanar (contained in the same 3D plane), instead these should be merged into one polygon."

It seems really strange, since [three.js geometries only work with triangular faces](https://threejs.org/docs/#api/en/core/Face3) (quads are cut in half). This is also how GLTF export works. In Blender I had a plain cube, with 6 faces, but the exported model has 12 faces.
![image](https://user-images.githubusercontent.com/39943985/100739458-61661f80-33d7-11eb-9d11-c197b87e20b8.png)
Also if instead of my imported geometry I pick the native BoxGeometry (which works just fine and collides without crashing), and log it, I'll find that it has indeed 12 faces.
![image](https://user-images.githubusercontent.com/39943985/100739652-ab4f0580-33d7-11eb-9dd5-ef322a15cb94.png)
I have no idea what to do with that. It sounds very very strange that cannon.js convex polyhedron can't have coplanar polygons, while three.js specifically work that way.

**Comment by Faber-smythe** [comment, marked as answer]
So…
**THE CAUSE** :
In the process of cuting quads into triangles, **Blender GLTF export multiplies the vertices**. This doesn’t bother the rendering of the meshes in Three.js, but **causes crashing deep into cannon-es script** when two of them **collide** with each other.

**THE FIX** :
The geometry can be cleaned up with [geometry.mergeVertices()](https://threejs.org/docs/#api/en/core/Geometry.mergeVertices). This caused the duplicated vertices to disappear, and the geometry.faces to be remapped onto the new geometry.vertices.

The [CodeSandBox ](https://codesandbox.io/s/convex-polyhedron-collision-hd1em?file=/src/App.js) here shows the issue and the fix.

**Reply by stockhuman** [reply]
Well done on troubleshooting this on your own! For the next person that comes around, note that some stuff like this [is mentioned within the demos](https://github.com/pmndrs/use-cannon/blob/576d7967935dbbfcd44b81347caab43487382702/examples/src/demos/ConvexPolyhedron.js#L14), where you might find other useful "recipes". :)
